pynq_composable/composable.py

```diff
@@ -476,141 +510,107 @@ def compose(self, cle_list: list) -> None:
-                                       self._relative_path(nextip._fullpath),
-                                       label=_edge_label(si[j], mi, gdebug))
+                        value = in_use_si[path]['si'][0]
+                        new_si = in_use_si[path]['si'][1:0]
+
+                        if len(new_si) > 0:
+                            in_use_si[path]['si'] = new_si
+
                         else:
-                            raise SystemError("IP: {} is already being used "
-                                              "in the provided pipeline. An IP"
-                                              " instance can only be used once"
-                                              .format(ip._fullpath))
+                            in_use_si.pop(path)
+                else:
+                    value = si[0]
 
+                if len(mi) > 1:
+                    if nextkey not in in_use_mi.keys():
+                        index = mi[0]
+                        in_use_mi[nextkey] = {'mi': mi[1:]}
+                    else:
+                        index = in_use_mi[nextkey]['mi'][0]
+                        new_mi = in_use_mi[nextkey]['mi'][1:0]
+                        if len(new_mi) > 0:
+                            in_use_mi[nextkey]['mi'] = new_mi
+                        else:
+                            in_use_mi.pop(nextkey)
+                else:
+                    index = mi[0]
+                if not np.where(switch_conf == value)[0].size:
+                    if switch_conf[index] != -1:
+                        raise SystemError("cannot meet pipeline "
+                                          "requirement of 2 input")
+                    switch_conf[index] = value
+                    graph.edge(self._relative_path(ip._fullpath),
+                               self._relative_path(nextip._fullpath),
+                               label=_edge_label(value, index, gdebug))
+                else:
+                    if max_si != max_mi:
+                        raise SystemError("cannot meet pipeline "
+                                          "requirement of 2 output")
+                    raise SystemError("IP: {} is already being used "
+                                      "in the provided pipeline. An IP"
+                                      " instance can only be used once"
+                                      .format(ip._fullpath))
+        if in_use_mi or in_use_si:
+            raise SystemError("Not all IPs within the pipeline "
+                              "were assigned. IP(s): {} {} "
+                              "are not connected correctly. "
+                              "Pipeline is invalid"
+                              .format(in_use_si, in_use_mi))
         if self._soft_reset and self._enable_soft_reset:
             self._soft_reset[0].write(1)
             self._soft_reset[0].write(0)
 
         self._configure_switch(switch_conf)
-
-        for idx, ip in enumerate(flat_list):
```

Comment:
This code cannot be removed. It needs to be refactor to use new flat_list

Reply:
Sorry I did not realize I had removed it!